src/dataflow/src/source/mod.rs

```diff
@@ -849,6 +853,221 @@ enum MessageProcessing {
+/// transactions submitted to the system. Rows can be inserted individually using the
+/// [insert](Timestamper::insert) and [delete](Timestamper::delete) methods or as part of a bigger
+/// transaction.
+///
+/// When a transaction is started using [start_tx](Timestamper::start_tx) the internal clock will be
+/// frozen and any subsequent rows will be timestamped with the exact same timestamp. The
+/// transaction is committed automatically as soon as the transaction object gets dropped.
+pub struct Timestamper {
+    inner: Arc<RwLock<Timestamp>>,
+    sender: EventSender,
+    tick_duration: Duration,
+}
+
+impl Timestamper {
+    fn new(sender: EventSender, tick_duration: Duration) -> Self {
+        let now = UNIX_EPOCH
+            .elapsed()
+            .expect("system clock before 1970")
+            .as_millis()
+            .try_into()
+            .expect("materialize has existed for more than 500M years");
+        Self {
+            inner: Arc::new(RwLock::new(now)),
+            sender,
+            tick_duration,
+        }
+    }
+
+    /// Start a transaction at a particular point in time. The timestamper will freeze its internal
+    /// clock while a transaction is active.
+    pub async fn start_tx<'a>(&'a self) -> SourceTransaction<'a> {
+        SourceTransaction {
+            timestamp: self.inner.read().await,
+            sender: &self.sender,
+        }
+    }
+
+    /// Record an insertion of a row
+    pub async fn insert(&self, row: Row) -> anyhow::Result<()> {
+        self.start_tx().await.insert(row).await
+    }
+
+    /// Record a deletion of a row
+    pub async fn delete(&self, row: Row) -> anyhow::Result<()> {
+        self.start_tx().await.delete(row).await
+    }
+
+    /// Records an error. After this method is called the source will permanently be in an errored
+    /// state
+    pub async fn error(&self, err: SourceError) -> anyhow::Result<()> {
+        let timestamp = self.inner.read().await;
+        self.sender
+            .send(Event::Message(*timestamp, Err(err)))
+            .await
+            .or_else(|_| Err(anyhow!("channel closed")))
+    }
+
+    /// Attempts to monotonically increase the current timestamp and provides a Progress message to
+    /// timely.
+    ///
+    /// This method will wait for all current transactions to commit before advancing the clock and
+    /// will cause any new requests for transactions to wait for the tick to complete before
+    /// starting.  This is due to the write-preferring behaviour of the tokio RwLock.
+    async fn tick(&self) -> anyhow::Result<()> {
+        tokio::time::sleep(self.tick_duration).await;
+        let mut timestamp = self.inner.write().await;
+        let now = UNIX_EPOCH
+            .elapsed()
+            .expect("system clock before 1970")
+            .as_millis()
+            .try_into()
+            .expect("materialize has existed for more than 500M years");
+        if *timestamp < now {
```

Comment:
the current code tries hard to pick timestamps aligned to the frequency interval which i think reduces the number of distinct timestamps going through the system for multiple source instances. idk if thats worth doing for now (also happy to have a TODO and do it later)

Reply:
I'll check if it's easy to do and if not leave a TODO. Out of curiosity, what is the relation of distinct timestamps and performance/memory?

Reply:
turns out it's a simple modulo, fixed!